### PR TITLE
[Trainer] Rename tokenizer to processor, add deprecation

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR updates the `tokenizer` argument which people can pass to the Trainer and TrainerCallback classes to `processor` instead. This allows for people to pass tokenizers, image processors, feature extractors or multimodal, which it will then automatically save along the model when training.

Follow-up of #29896

To do:

- [ ] perhaps use `preprocessor` instead of `processor` to avoid confusion with multimodal processors like `CLIPProcessor`?
- [x] add deprecation message for tokenizer attribute

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -360,14 +363,24 @@ def __init__(
...
+            Union[PreTrainedTokenizerBase, BaseImageProcessor, SequenceFeatureExtractor, ProcessorMixin]
+        ] = None,
         model_init: Optional[Callable[[], PreTrainedModel]] = None,
         compute_metrics: Optional[Callable[[EvalPrediction], Dict]] = None,
         callbacks: Optional[List[TrainerCallback]] = None,
         optimizers: Tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR] = (None, None),
         preprocess_logits_for_metrics: Optional[Callable[[torch.Tensor, torch.Tensor], torch.Tensor]] = None,
+        tokenizer: Optional[PreTrainedTokenizerBase] = None,
     ):
+        if tokenizer is not None and not isinstance(tokenizer, PreTrainedTokenizerBase):
```
Comment (amyeroberts) [inline src/transformers/trainer.py:376]: We shouldn't check against the type here - we want people to use `processor` even if they're using the old API of `tokenizer=image_processor`

File: src/transformers/trainer_callback.py [hunk 2]
```diff
@@ -309,13 +309,20 @@ def on_prediction_step(self, args: TrainingArguments, state: TrainerState, contr
 class CallbackHandler(TrainerCallback):
     """Internal class that just calls the list of callbacks in order."""
 
-    def __init__(self, callbacks, model, tokenizer, image_processor, optimizer, lr_scheduler):
+    def __init__(self, callbacks, model, processor, optimizer, lr_scheduler, tokenizer=None):
+        if tokenizer is not None and not isinstance(tokenizer, PreTrainedTokenizerBase):
```
Comment (amyeroberts) [inline src/transformers/trainer_callback.py:313]: Same here

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -668,6 +684,14 @@ def __init__(
             num_devices = xr.global_runtime_device_count()
             xs.set_global_mesh(xs.Mesh(np.array(range(num_devices)), (num_devices, 1), axis_names=("fsdp", "tensor")))
 
+    @property
+    def tokenizer(self):
+        warnings.warn(
+            "The 'tokenizer' attribute is deprecated and will be removed in a future version. Use `processor` instead",
```
Comment (amyeroberts) [inline src/transformers/trainer.py:690]: Version this will be deprecated in should be specified here

**Review by amyeroberts — COMMENTED** [review]
Thanks for updating - looks a lot better with `processor`! 

* Other examples which have `tokenizer=tokenizer` will also need to be updated if they exist
* Question over the deprecation cycle - v5 means it won't be remove in a long time - but as this is a fairly large change from a commonly used API I think it's OK

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -360,14 +363,24 @@ def __init__(
...
-        image_processor: Optional["BaseImageProcessor"] = None,
+        processor: Optional[
+            Union[PreTrainedTokenizerBase, BaseImageProcessor, SequenceFeatureExtractor, ProcessorMixin]
+        ] = None,
         model_init: Optional[Callable[[], PreTrainedModel]] = None,
         compute_metrics: Optional[Callable[[EvalPrediction], Dict]] = None,
         callbacks: Optional[List[TrainerCallback]] = None,
         optimizers: Tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR] = (None, None),
         preprocess_logits_for_metrics: Optional[Callable[[torch.Tensor, torch.Tensor], torch.Tensor]] = None,
+        tokenizer: Optional[PreTrainedTokenizerBase] = None,
```
Comment (younesbelkada) [inline src/transformers/trainer.py:374]: I think we should be careful when manipulating the init signature, see: https://github.com/huggingface/transformers/pull/30126 - for new args we should put them at the end

**Comment by NielsRogge** [comment]
Thanks, I opened #30129 as a better quick fix cause there are a huge amount of files having `tokenizer=tokenizer`.

The update regarding

**Comment by amyeroberts** [comment]
@NielsRogge Although #30129 proposes a solution for the default data collator, we should still add this update.  `tokenizer=image_processor` is a confusing API (as highlighted by a few issues from users) and considering we have more and more multimodal, audio and vision models increasingly out-of-date

Replacing `tokenizer=tokenizer` with `processor=tokenizer` might touch a lot of files but I'd expect to be fairly straightforward with some grep and find/replace.

**Comment by sanchit-gandhi** [comment]
Thanks for pointing me in the direction of this PR @amyeroberts! I personally don't think replacing `tokenizer` by `processor` is a good idea in the context of NLP users: many NLP users won't know the notion of a `processor`, since they only use the `tokenizer` to pre- and post-process their data. To me, having them set `processor=tokenizer` is just as confusing as having multimodal users set `tokenizer=processor`.

I would be more in favour of a design where allow NLP users to pass `tokenizer=tokenizer`, and multimodal users to pass `processor=processor`, as suggested in https://github.com/huggingface/transformers/pull/30864#discussion_r1603659097 -> to me, this is the cleanest and most intuitive API. Yes, it adds some complexity under-the-hood in the Trainer, since we now have to handle both object, but this is a valid burden to make the API as intuitive as possible for the user.

**Comment by amyeroberts** [comment]
@sanchit-gandhi I think what's being suggested is actually a flag more like `preprocessor` or `processing_class` which would replace `tokenizer`. As Trainer is such a commonly used object, in reality we're unlikely to fully remove or deprecate `tokenizer` anytime soon, but I don't think this means we have to add a whole new argument for every processing class.

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -360,14 +366,24 @@ def __init__(
...
         compute_metrics: Optional[Callable[[EvalPrediction], Dict]] = None,
         callbacks: Optional[List[TrainerCallback]] = None,
         optimizers: Tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR] = (None, None),
         preprocess_logits_for_metrics: Optional[Callable[[torch.Tensor, torch.Tensor], torch.Tensor]] = None,
-        image_processor: Optional["BaseImageProcessor"] = None,
+        tokenizer: Optional[PreTrainedTokenizerBase] = None,
     ):
+        if tokenizer is not None:
+            warnings.warn(
+                "The `tokenizer` argument is deprecated and will be removed in v5 of Transformers. You can use `processor` "
```
Comment (sanchit-gandhi) [inline src/transformers/trainer.py:381]: I believe this PR is proposing deprecating `tokenizer` @amyeroberts! I think `preprocessor` is limited in the sense that we don't just want to save the pre-processing class (e.g. image processor or feature extractor), but also the post-processing class (e.g. the tokenizer) => for this reason I think passing the `processor` is cleanest here (for multimodal models)

Comment (amyeroberts) [reply src/transformers/trainer.py:381]: Ah, yes. Perhaps I wasn't clear. What I meant was we would introduce an new argument, which should be used in preference to `tokenizer` which would accept all of the processing class. However we wouldn't fully deprecate `tokenizer` for a long time, and still allow it to be used

Comment (sanchit-gandhi) [reply src/transformers/trainer.py:381]: Great thanks for the clarification - on the same page here! Happy to update the PR #30864 accordingly, unless you want to see this one to completion @NielsRogge?

**Pull request closed without merging**